pkg/storage/gcs/gcp_labels_tags.go

```diff
@@ -35,3 +79,220 @@ func getUserLabels(infraLister configlisters.InfrastructureLister) (map[string]s
 	}
 	return labels, nil
 }
+
+// newLimiter returns token bucket based request rate limiter after initializing
+// the passed values for limit, burst(or token bucket) size. If opted for emptyBucket
+// all initial tokens are reserved for the first burst.
+func newLimiter(limit, burst int, emptyBucket bool) *rate.Limiter {
+	limiter := rate.NewLimiter(rate.Every(time.Second/time.Duration(limit)), burst)
+
+	if emptyBucket {
+		limiter.AllowN(time.Now(), burst)
+	}
+
+	return limiter
+}
+
+// toTagValueList converts the tags to an array containing tagValues
+// NamespacedNames.
+func toTagValueList(tags []configv1.GCPResourceTag) []string {
+	if len(tags) <= 0 {
+		return nil
+	}
+
+	list := make([]string, 0, len(tags))
+	for _, tag := range tags {
+		t := fmt.Sprintf("%s/%s/%s", tag.ParentID, tag.Key, tag.Value)
+		list = append(list, t)
+	}
+	return list
+}
+
+// getInfraResourceTagsList returns the user-defined tags present in the
+// status sub-resource of Infrastructure.
+func getInfraResourceTagsList(platformStatus *configv1.PlatformStatus) []configv1.GCPResourceTag {
+	if platformStatus != nil && platformStatus.GCP != nil && platformStatus.GCP.ResourceTags != nil {
+		return platformStatus.GCP.ResourceTags
+	}
+	klog.V(1).Infof("getInfraResourceTagsList: user-defined tag list is empty")
+	return nil
+}
+
+// getTagsList returns the list of tags to apply on the resources.
+func getTagsList(platformStatus *configv1.PlatformStatus) []string {
+	return toTagValueList(getInfraResourceTagsList(platformStatus))
+}
+
+// getFilteredTagList returns the list of tags to apply on the resources after
+// filtering the tags already existing on a given resource.
+func getFilteredTagList(ctx context.Context, platformStatus *configv1.PlatformStatus, client *rscmgr.TagBindingsClient, parent string) []string {
+	return filterTagList(ctx, client, parent, getTagsList(platformStatus))
+}
+
+// filterTagList returns the filtered list of tags to apply on the resources.
+func filterTagList(ctx context.Context, client *rscmgr.TagBindingsClient, parent string, tagList []string) []string {
+	dupTags := make(map[string]bool, len(tagList))
+	for _, k := range tagList {
+		dupTags[k] = false
+	}
+
+	listBindingsReq := &rscmgrpb.ListEffectiveTagsRequest{
+		Parent: parent,
+	}
+	bindings := client.ListEffectiveTags(ctx, listBindingsReq)
+	// a resource can have a maximum of 50 tags attached to it. Will iterate
+	// for 50 times in the worst case scenario of none of the break conditions are met.
+	for i := 0; i < 50; i++ {
```

Comment:
I am little hesitant to use unconditional loops in case the break conditions aren't met and it goes into infinite loop.
My idea here was to reduce the number of tag write API calls by looking for existing tags and skipping them, since it has per minute quota restriction. 

If the max tags number is increased, we would be able skip those in the first 50(which will not meet the purpose of the code here) but the logic [here](https://github.com/openshift/cluster-image-registry-operator/pull/873/files/33417f06aa5f83823ec346cf4b1aaaa1ed060cd5..6f8f74e620813410862c9fcc292fde8d9aede33c#diff-6e6f5be6182d62cd1ad9867497387a3c45de233050881e4479396cea0652d4a5R279) will handle existing tags error which weren't skipped. And this would result in additional write calls for already existing tags, but still work as expected, I think. 

Please let me know your thoughts.

Reply:
Okay so you're saying that even if the max number of tags per resource increase and this code can no longer get all of the tags because of that - it would be okay because the filtering done in this function is an optimization to reduce the number of api calls when trying to tag resources. Got it.

In that case, let's just add this explanation as a comment to the code. WDYT?

Reply:
suggestion incorporated, updated the comment.